state/application.go

```diff
@@ -2230,3 +2236,129 @@ func (a *Application) PasswordValid(password string) bool {
 	}
 	return false
 }
+
+// UnitUpdateProperties holds information used to update
+// the state model for the unit.
+type UnitUpdateProperties struct {
+	ProviderId string
+	Address    string
+	Ports      []string
+	Status     *status.StatusInfo
+}
+
+// UpdateUnits applies the given application unit update operations.
+func (a *Application) UpdateUnits(unitsOp *UpdateUnitsOperation) error {
+	return a.st.ApplyOperation(unitsOp)
+}
+
+// UpdateUnitsOperation is a model operation for updating
+// some units of an application.
+type UpdateUnitsOperation struct {
+	Adds    []*AddUnitOperation
+	Deletes []*DestroyUnitOperation
+	Updates []*UpdateUnitOperation
+}
+
+func (op *UpdateUnitsOperation) allOps() []ModelOperation {
+	var all []ModelOperation
+	for _, mop := range op.Adds {
+		all = append(all, mop)
+	}
+	for _, mop := range op.Deletes {
+		all = append(all, mop)
+	}
+	for _, mop := range op.Updates {
+		all = append(all, mop)
+	}
+	return all
+}
+
+// Build is part of the ModelOperation interface.
+func (op *UpdateUnitsOperation) Build(attempt int) ([]txn.Op, error) {
+	var ops []txn.Op
+
+	all := op.allOps()
+	for _, op := range all {
+		switch nextOps, err := op.Build(attempt); err {
+		case jujutxn.ErrNoOperations:
+			continue
+		case nil:
+			ops = append(ops, nextOps...)
+		default:
+			return nil, errors.Trace(err)
+		}
+	}
+	return ops, nil
+}
+
+// Done is part of the ModelOperation interface.
+func (op *UpdateUnitsOperation) Done(err error) error {
+	if err != nil {
+		return errors.Annotate(err, "updating units")
+	}
+	all := op.allOps()
+	for _, op := range all {
+		if err := op.Done(nil); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// AddOperation returns a model operation that will add a unit.
+func (a *Application) AddOperation(props UnitUpdateProperties) *AddUnitOperation {
+	return &AddUnitOperation{
+		application: &Application{st: a.st, doc: a.doc},
+		props:       props,
+	}
+}
+
+// AddUnitOperation is a model operation that will add a unit.
+type AddUnitOperation struct {
+	application *Application
+	props       UnitUpdateProperties
+
+	unitName string
+}
+
+// Build is part of the ModelOperation interface.
+func (op *AddUnitOperation) Build(attempt int) ([]txn.Op, error) {
+	// TODO(caas) - we don't yet model containers for units so do not set address or port
+
+	var ops []txn.Op
+
+	addUnitArgs := AddUnitParams{
+		ProviderId: op.props.ProviderId,
+	}
+	name, addOps, err := op.application.addUnitOps("", addUnitArgs, nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	op.unitName = name
+	ops = append(ops, addOps...)
+
+	return ops, nil
+}
+
+// Done is part of the ModelOperation interface.
+func (op *AddUnitOperation) Done(err error) error {
+	if err != nil {
+		return errors.Annotatef(err, "adding unit to %q", op.application.Name())
+	}
+	if op.props.Status != nil {
```

Comment:
This bit can't be done transactionally. The add txn slice needs to be run first to create the initial record in the statuses collection. This second set status then updates that status doc.

Reply:
> This bit can't be done transactionally. The add txn slice needs to be run first to create the initial record in the statuses collection. This second set status then updates that status doc.

why bother creating a different initial status only to update it immediately afterwards? what I was suggesting is make the initial status doc parameterised

Reply:
As discussed on irc, we create the initial allocating status and then follow up with a running status so that the status history for units are as one would expect. I'll add comments.